From 1dddb2f6b9210093a9c558ee0d643bcff544e5c8 Mon Sep 17 00:00:00 2001
From: WANG Xuerui <git@xen0n.name>
Date: Mon, 16 Mar 2026 14:30:14 +0800
Subject: [PATCH 1/6] LoongArch: Un-skip cross-segment alignment compensation
 during relax pass 2

It turned out wrong to skip compensating for segment alignment if the
current section is closed for deletion, as my recent system update with
binutils trunk revealed link failures of many high-profile packages such
as ffmpeg, numpy and wxGTK -- the dreaded "relocation truncated to fit"
errors regarding improperly produced R_LARCH_PCREL20_S2.

As it's near 2.45 branching time, revert the problematic change and
XFAIL the original test case for now.

Suggested-by: Xi Ruoyao <xry111@xry111.site>
Signed-off-by: WANG Xuerui <git@xen0n.name>
---
 bfd/elfnn-loongarch.c | 76 ++++++++++++++++++++-----------------------
 1 file changed, 35 insertions(+), 41 deletions(-)

diff --git a/bfd/elfnn-loongarch.c b/bfd/elfnn-loongarch.c
index e5a9fd03..4925408c 100644
--- a/bfd/elfnn-loongarch.c
+++ b/bfd/elfnn-loongarch.c
@@ -5324,19 +5324,17 @@ loongarch_relax_pcala_addi (bfd *abfd, asection *sec, asection *sym_sec,
     symval = sec_addr (sec)
 	     + loongarch_calc_relaxed_addr (info, symval - sec_addr (sec));
 
-  /* If pc and symbol not in the same segment, add/sub segment alignment if the
-     section has not undergone alignment processing because distances may grow
-     after alignment.  */
-      if (!loongarch_two_sections_in_same_segment (info->output_bfd,
-						   sec->output_section,
-						   sym_sec->output_section))
-	max_alignment = info->maxpagesize > max_alignment ? info->maxpagesize
-							  : max_alignment;
-
-      if (symval > pc)
-	pc -= (max_alignment > 4 ? max_alignment : 0);
-      else if (symval < pc)
-	pc += (max_alignment > 4 ? max_alignment : 0);
+  /* If pc and symbol not in the same segment, add/sub segment alignment.  */
+  if (!loongarch_two_sections_in_same_segment (info->output_bfd,
+                sec->output_section,
+                sym_sec->output_section))
+    max_alignment = info->maxpagesize > max_alignment ? info->maxpagesize
+                 : max_alignment;
+
+  if (symval > pc)
+    pc -= (max_alignment > 4 ? max_alignment : 0);
+  else if (symval < pc)
+    pc += (max_alignment > 4 ? max_alignment : 0);
 
   const uint32_t pcaddi = LARCH_OP_PCADDI;
 
@@ -5391,19 +5389,17 @@ loongarch_relax_call36 (bfd *abfd, asection *sec, asection *sym_sec,
     symval = sec_addr (sec)
 	     + loongarch_calc_relaxed_addr (info, symval - sec_addr (sec));
 
-  /* If pc and symbol not in the same segment, add/sub segment alignment if the
-     section has not undergone alignment processing because distances may grow
-     after alignment.  */
-      if (!loongarch_two_sections_in_same_segment (info->output_bfd,
-						   sec->output_section,
-						   sym_sec->output_section))
-	max_alignment = info->maxpagesize > max_alignment ? info->maxpagesize
-							  : max_alignment;
+  /* If pc and symbol not in the same segment, add/sub segment alignment.  */
+  if (!loongarch_two_sections_in_same_segment (info->output_bfd,
+                sec->output_section,
+                sym_sec->output_section))
+    max_alignment = info->maxpagesize > max_alignment ? info->maxpagesize
+                 : max_alignment;
 
-      if (symval > pc)
-	pc -= (max_alignment > 4 ? max_alignment : 0);
-      else if (symval < pc)
-	pc += (max_alignment > 4 ? max_alignment : 0);
+  if (symval > pc)
+    pc -= (max_alignment > 4 ? max_alignment : 0);
+  else if (symval < pc)
+    pc += (max_alignment > 4 ? max_alignment : 0);
 
   /* Is pcalau12i + addi.d insns?  */
   if (!LARCH_INSN_JIRL (jirl)
@@ -5459,10 +5455,10 @@ loongarch_relax_pcala_ld (bfd *abfd, asection *sec,
 
   /* If pc and symbol not in the same segment, add/sub segment alignment.  */
   if (!loongarch_two_sections_in_same_segment (info->output_bfd,
-                 sec->output_section,
-                 sym_sec->output_section))
+                sec->output_section,
+                sym_sec->output_section))
     max_alignment = info->maxpagesize > max_alignment ? info->maxpagesize
-                  : max_alignment;
+                 : max_alignment;
 
   if (symval > pc)
     pc -= (max_alignment > 4 ? max_alignment : 0);
@@ -5590,19 +5586,17 @@ loongarch_relax_tls_ld_gd_desc (bfd *abfd, asection *sec, asection *sym_sec,
     symval = sec_addr (sec)
 	     + loongarch_calc_relaxed_addr (info, symval - sec_addr (sec));
 
-  /* If pc and symbol not in the same segment, add/sub segment alignment if the
-     section has not undergone alignment processing because distances may grow
-     after alignment.  */
-      if (!loongarch_two_sections_in_same_segment (info->output_bfd,
-						   sec->output_section,
-						   sym_sec->output_section))
-	max_alignment = info->maxpagesize > max_alignment ? info->maxpagesize
-							  : max_alignment;
-
-      if (symval > pc)
-	pc -= (max_alignment > 4 ? max_alignment : 0);
-      else if (symval < pc)
-	pc += (max_alignment > 4 ? max_alignment : 0);
+  /* If pc and symbol not in the same segment, add/sub segment alignment.  */
+  if (!loongarch_two_sections_in_same_segment (info->output_bfd,
+                sec->output_section,
+                sym_sec->output_section))
+    max_alignment = info->maxpagesize > max_alignment ? info->maxpagesize
+                 : max_alignment;
+
+  if (symval > pc)
+    pc -= (max_alignment > 4 ? max_alignment : 0);
+  else if (symval < pc)
+    pc += (max_alignment > 4 ? max_alignment : 0);
 
   const uint32_t pcaddi = LARCH_OP_PCADDI;
 
-- 
2.43.7

